sklearn/manifold/tests/test_spectral_embedding.py

```diff
@@ -117,33 +126,64 @@ def test_spectral_embedding_two_components(seed=36):
     true_label[0:n_sample] = 1
 
     se_precomp = SpectralEmbedding(
-        n_components=1, affinity="precomputed", random_state=np.random.RandomState(seed)
+        n_components=1,
+        affinity="precomputed",
+        random_state=np.random.RandomState(seed),
+        eigen_solver=eigen_solver,
     )
-    embedded_coordinate = se_precomp.fit_transform(affinity)
-    # Some numpy versions are touchy with types
-    embedded_coordinate = se_precomp.fit_transform(affinity.astype(np.float32))
-    # thresholding on the first components using 0.
-    label_ = np.array(embedded_coordinate.ravel() < 0, dtype="float")
-    assert normalized_mutual_info_score(true_label, label_) == pytest.approx(1.0)
+    for dtype in [np.float32, np.float64]:
+        if eigen_solver == "amg" and not amg_loaded:
+            # with pytest.raises(
+            #     ValueError, match="The eigen_solver was set to 'amg', but"
+            # ):
+            #     embedded_coordinate = se_precomp.fit_transform(
+            #         affinity.astype(dtype)
+            #     )
+            pass
+        else:
+            embedded_coordinate = se_precomp.fit_transform(affinity.astype(dtype))
+            # thresholding on the first components using 0.
+            label_ = np.array(embedded_coordinate.ravel() < 0, dtype="float")
+            assert normalized_mutual_info_score(true_label, label_) == pytest.approx(
+                1.0
+            )
 
 
 @pytest.mark.parametrize("X", [S, sparse.csr_matrix(S)], ids=["dense", "sparse"])
-def test_spectral_embedding_precomputed_affinity(X, seed=36):
+@pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg"))
+@pytest.mark.parametrize("dtype", (np.float32, np.float64))
+def test_spectral_embedding_precomputed_affinity(X, eigen_solver, dtype):
+    seed = 36
     # Test spectral embedding with precomputed kernel
     gamma = 1.0
     se_precomp = SpectralEmbedding(
-        n_components=2, affinity="precomputed", random_state=np.random.RandomState(seed)
+        n_components=2,
+        affinity="precomputed",
+        random_state=np.random.RandomState(seed),
+        eigen_solver=eigen_solver,
     )
     se_rbf = SpectralEmbedding(
         n_components=2,
         affinity="rbf",
         gamma=gamma,
         random_state=np.random.RandomState(seed),
+        eigen_solver=eigen_solver,
     )
-    embed_precomp = se_precomp.fit_transform(rbf_kernel(X, gamma=gamma))
-    embed_rbf = se_rbf.fit_transform(X)
-    assert_array_almost_equal(se_precomp.affinity_matrix_, se_rbf.affinity_matrix_)
-    _assert_equal_with_sign_flipping(embed_precomp, embed_rbf, 0.05)
+    if eigen_solver == "amg" and not amg_loaded:
+        # with pytest.raises(
+        #     ValueError, match="The eigen_solver was set to 'amg', but"
+        # ):
+        #     embedded_coordinate = se_precomp.fit_transform(
+        #         affinity.astype(dtype)
+        #     )
+        pass
```

Comment:
Either we need to enable this check for the `ValueError` and its message or just skip this test with `pytest.importorskip("pyamg")`.

Note that sometimes it's necessary to wrap the expected error message with `re.escape`, e.g.:

```python
match=re.escape("The eigen_solver was set to 'amg', but")
```

since the `match` parameter expects a regular expression from the `re` module.

Reply:
Thanks!
I have found another way at https://code-maven.com/pytest-assert-error-reporting that separates assertions of  raising the error and the content  of the error message, which is nicer. Testing it now.